src/snowflake/connector/s3_storage_client.py

```diff
@@ -87,84 +92,145 @@ def __init__(
-            res += f"/{bucket_name}"
-            if request_uri:
-                res += "/" + request_uri
-        else:
-            # for GET operations without a bucket name
-            res += "/"
-        if subresource:
-            res += "?"
-            keys = sorted(subresource.keys())
-            res += (
-                keys[0]
-                if subresource[keys[0]] is None
-                else f"{keys[0]}={subresource[keys[0]]}"
-            )
-            for k in keys[1:]:
-                query_str = k if subresource[k] is None else f"{k}={subresource[k]}"
-                res += f"&{query_str}"
-        return res
+        """Convenience function to build the query part of a URL from key-value pairs.
+
+        It filters out empty strings from the key, value pairs.
+        """
+        return "&".join(["=".join(filter(bool, e)) for e in query_parts])
 
     @staticmethod
-    def construct_canonicalized_headers(
+    def _construct_canonicalized_and_signed_headers(
         headers: Dict[str, Union[str, List[str]]]
-    ) -> str:
-        _res = sorted([k.lower(), v] for k, v in headers.items())
+    ) -> Tuple[str, str]:
+        """Construct canonical headers as per AWS specs, returns the signed headers too.
+
+        Does not support sorting by values in case the keys are the same, don't send
+        in duplicate keys, but this is not possible with a dictionary anyways.
+        """
         res = []
+        low_key_dict = {k.lower(): v for k, v in headers.items()}
+        sorted_headers = sorted(low_key_dict.keys())
+        _res = [(k, low_key_dict[k]) for k in sorted_headers]
 
-        for i in range(len(_res)):
-            k, v = _res[i]
+        for k, v in _res:
             # if value is a list, convert to string delimited by comma
             if isinstance(v, list):
                 v = ",".join(v)
             # if multiline header, replace withs space
             k = k.replace("\n", " ")
-            res.append(k.rstrip() + ":" + v.lstrip())
+            res.append(k.strip() + ":" + RE_MULTIPLE_SPACES.sub(" ", v.strip()))
 
         ans = "\n".join(res)
         if ans:
-            ans = ans + "\n"
+            ans += "\n"
 
-        return ans
+        return ans, ";".join(sorted_headers)
 
     @staticmethod
-    def _construct_string_to_sign(
+    def _construct_canonical_request_and_signed_headers(
         verb: str,
-        canonicalized_element: str,
-        canonicalized_headers: str,
-        amzdate: str,
-        content_md5: str = "",
-        content_type: str = "",
-    ) -> bytes:
-        res = verb + "\n" + content_md5 + "\n" + content_type + "\n"
-        res += amzdate + "\n" + canonicalized_headers + canonicalized_element
-        return res.encode("UTF-8")
+        canonical_uri_parameter: str,
+        query_parts: Dict[str, str],
+        canonical_headers: Optional[Dict[str, Union[str, List[str]]]] = None,
+        payload_hash: str = "",
+    ) -> Tuple[str, str]:
```

Comment:
Essentially this function does step 1 from https://github.com/snowflakedb/snowflake-connector-python/pull/846#issuecomment-914643727 signed headers are basically computed in here and they will be necessary for part of the string to sign generation later on.